examples/benchmark.py

```diff
@@ -31,183 +124,469 @@ def downloadAmberSuite():
+        The System to serialize
+    integrator : openmm.Integrator
+        The Integrator to serialize
+    state : openmm.State
+        The State to serialize
+    coredata : dict
+        Core parameters to serialize to XML
+    metadata : dict
+        Metadata to dump to YAML
+    """
+    import os
+    os.makedirs(directory, exist_ok=True)
+    import openmm
+    def serialize(obj, filename):
+        if filename.endswith('.bz2'):
+            import bz2
+            with bz2.open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+        elif filename.endswith('.gz'):
+            import gzip
+            with gzip.open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+        else:
+            with open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+
+    serialize(system, os.path.join(directory, 'system.xml.bz2'))
+    serialize(integrator, os.path.join(directory, 'integrator.xml.bz2'))
+    serialize(state, os.path.join(directory, 'state.xml.bz2'))
+
+    with open(os.path.join(directory, 'core.xml'), 'wt') as outfile:
+        outfile.write('<config>\n')
+        for key, value in coredata.items():
+            outfile.write(f'    <{key} v="{value}"/>\n')
+        outfile.write('</config>\n')
+
+    import yaml
+    with open(os.path.join(directory, 'metadata.yaml'), 'wt') as outfile:
+        outfile.write(yaml.dump(metadata))
+
+def runOneTest(testName, options):
+    """Perform a single benchmarking simulation."""
+
+    # Return if we do not support the requested precision
+    if ((options.platform == 'Reference') and (options.precision != 'double')) \
+       or ((options.platform == 'CPU') and (options.precision != 'mixed')):
+        return
+
+    system, positions = retrieveTestSystem(testName, pme_cutoff=options.pme_cutoff, bond_constraints=options.bond_constraints, polarization=options.polarization, epsilon=options.epsilon)
+
+    test_result = dict()
+    test_result['test'] = testName
+
+    explicit = (testName not in ('gbsa', 'amoebagk'))
+    amoeba = (testName in ('amoebagk', 'amoebapme'))
+    apoa1 = testName.startswith('apoa1')
+    amber = (testName.startswith('amber'))
+    hydrogenMass = None
+
+    if amoeba:
+        test_result['epsilon'] = options.epsilon
+    elif (testName in ['pme', 'ljpme']) or testName.startswith('amber'):
+        test_result['cutoff'] = options.pme_cutoff
+
+    if amoeba:
+        test_result['constraints'] = 'None'
+        test_result['hydrogen_mass'] = '1'
+    elif options.bond_constraints == 'hbonds':
+        test_result['constraints'] = 'HBonds'
+        test_result['hydrogen_mass'] = '1.5'
+    elif options.bond_constraints == 'allbonds':
+        test_result['constraints'] = 'AllBonds'
+        test_result['hydrogen_mass'] = '4'
+    else:
+        raise ValueError(f'Unknown bond_constraints: {bond_constraints}')
```

Comment:
This duplicates logic that's already present in `retrieveTestSystem()`.  It would be better if we could unify them.  Otherwise there's a high risk that someone will modify one and not realize they have to modify the other as well.

Reply:
Fixed, but it would be better if there was a simpler, more modular way to register tests and their permitted values for these combinations of options.